pkg/scheduler/eventhandlers/reservation_handler.go

```diff
@@ -0,0 +1,132 @@
+limitations under the License.
+*/
+
+package eventhandlers
+
+import (
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/frameworkext"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/reservation"
+)
+
+var assignedPodDelete = framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete, Label: "AssignedPodDelete"}
+
+func AddScheduleEventHandler(sched *scheduler.Scheduler, extendedHandle frameworkext.ExtendedHandle) {
+	reservationInformer := extendedHandle.KoordinatorSharedInformerFactory().Scheduling().V1alpha1().Reservations().Informer()
+	reservationInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			r, ok := obj.(*schedulingv1alpha1.Reservation)
+			if ok {
+				addReservation(sched, r)
+			} else {
+				klog.Errorf("reservation informer add func parse obj failed, obj %T", obj)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldR, oldOK := oldObj.(*schedulingv1alpha1.Reservation)
+			newR, newOK := newObj.(*schedulingv1alpha1.Reservation)
+			if !oldOK || !newOK {
+				klog.Errorf("unable to convert object to *schedulingv1alpha1.Reservation, old %T, new %T", oldObj, newObj)
+				return
+			}
+			updateReservation(sched, oldR, newR)
+		},
+		DeleteFunc: func(obj interface{}) {
+			switch t := obj.(type) {
+			case *schedulingv1alpha1.Reservation:
+				deleteReservation(sched, t)
+				return
+			case cache.DeletedFinalStateUnknown:
+				r, ok := t.Obj.(*schedulingv1alpha1.Reservation)
+				if ok {
+					deleteReservation(sched, r)
+					return
+				}
+			}
+			klog.Errorf("reservation informer delete func parse obj failed, obj %T", obj)
+		},
+	})
+}
+
+func addReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to add reservation for invalid reservation %v", r)
+		return
+	}
+	reservePod := reservation.GetReservePod(r)
+	// TBD: should the reservation with nodeName specified enqueue?
+	if reservePod.Status.Phase != schedulingv1alpha1.ReasonReservationExpired {
+		err := sched.SchedulingQueue.Add(reservePod)
+		if err != nil {
+			klog.Errorf("failed to add reserve pod into scheduling queue, reservation %v, err: %v", reservation.GetReservationNamespacedName(r), err)
+		}
+		klog.V(4).Infof("add reservation %v", reservation.GetReservationNamespacedName(r))
+	}
+}
+
+func updateReservation(sched *scheduler.Scheduler, oldR, newR *schedulingv1alpha1.Reservation) {
+	if oldR == nil || newR == nil {
+		klog.Errorf("failed to update reservation for invalid reservation old %v, new %v", oldR, newR)
+		return
+	}
+	// only react to status.phase change
+	if oldR.Status.Phase == newR.Status.Phase {
```

Comment:
If the users modify the ReservationSpec, what should be done?  Or do we allow users to modify the ReservationSpec?

Reply:
Please update the design proposal.